Add WaterDrop integration (i.e.: trace messages produced by Karafka)

Drowze opened the pull request:
**What does this PR do?**

Adds support for WaterDrop (i.e. Karafka producers). Fixes #4868

**Motivation:**
We use Datadog and Karafka. For a long time we've been using a custom patch to have some distributed tracing support, but since @nvh0412 got the ball rolling by [adding traces for Karafka consumers](https://github.com/DataDog/dd-trace-rb/pull/4147), I've decided to continue the work there and also add proper tracing support for Karafka producers.
Hope we can get an ever-improving good support for Karafka in Datadog! 😄 

_(Thanks @mensfeld for the amazing Kafka framework!)_

**Change log entry**
<!-- If this is a customer-visible change, a brief summary to be placed
into the change log.
If you are not a Datadog employee, you can skip this section
and it will be filled or deleted during PR review. -->

**Additional Notes:**
Although WaterDrop is part of the Karafka framework, it can be used in isolation (which is handy if the application doesn't need any message consumer).

**How to test the change?**
See:
* [Karafka's getting started](https://karafka.io/docs/Getting-Started/) (`Karafka.producer` is a WaterDrop instance)
* [WaterDrop's getting started](https://karafka.io/docs/WaterDrop-Getting-Started/) (for using WaterDrop without Karafka)

<!-- Unsure? Have a question? Request a review! -->

Drowze committed:
Add WaterDrop integration

Drowze committed:
Fix thread leak warnings on Karafka tests

Drowze committed:
Add Karafka framework, automatically activating WaterDrop

Drowze committed:
Add WaterDrop to test matrix & docs

Drowze committed:
Update docs/GettingStarted.md

Co-authored-by: Esther Kim <esther.kim@datadoghq.com>

Drowze commented on docs/GettingStarted.md:
```diff
@@ -1158,10 +1158,32 @@ end
+
+The WaterDrop integration provides tracing of the `waterdrop` gem (a dependency of `karafka`, but also can be used standalone).
+
+This integration is automatically activated with the Karafka framework. If you're not using Karafka (e.g.: your app _produce_ messages for a different app _consume_), enable it through `Datadog.configure`:
```
You got it right. Your version looks much better than mine, thanks for the suggestion 😄  (just comitted it)

Drowze committed:
Fix karafka framework patch

Drowze committed:
Do not add partition tag if its missing

marcotc commented:
Thank you, @Drowze! We'll take a look in the next few days.

mensfeld commented:
I think we oculd tackle this by adding a process wide monitor that would react on builder events.

mensfeld commented:
> They just don't have a way to automatically add some middleware to all new instances.

tadam https://github.com/karafka/waterdrop/pull/680

Drowze committed:
WaterDrop: hook onto producer.configured

NOTE: producer.configured is a new feature, which should be present on
WaterDrop 2.8.8 (it's only present in the 2.8.8.rc1 for now).

vpellan commented on appraisal/ruby-3.0.rb:
```diff
@@ -97,12 +97,10 @@
 build_coverage_matrix('mongo', min: '2.1.0')
 build_coverage_matrix('dalli', [2])
+build_coverage_matrix('karafka', min: '2.3.0')
+build_coverage_matrix('waterdrop', min: '2.8.8.rc1')
```
`waterdrop` 2.8.8.rc1 specifies that it only supports Ruby 3.1 and above, which makes the CI fails for this file.

vpellan commented on appraisal/ruby-3.0.rb:
```diff
@@ -97,12 +97,10 @@
 build_coverage_matrix('rest-client')
 build_coverage_matrix('mongo', min: '2.1.0')
 build_coverage_matrix('dalli', [2])
+build_coverage_matrix('karafka', min: '2.3.0')
```
`karafka` 2.5.1 requires Ruby 3.2 or above, so we cannot have a 'latest' appraisal, only the min one. Looking at the karafka changelog, it seems that 2.5.1 was not the first version to not support Ruby 3.0 through transitive dependencies, although the gemspec said 3.0 was compatible, which is probably why there was a `karafka-min` appraise lower.

This should still be okay for Ruby 3.1 as long as we don't update the lockfiles, we can do that on our end.

vpellan commented on appraisal/ruby-3.1.rb:
```diff
@@ -98,6 +98,7 @@
 build_coverage_matrix('mongo', min: '2.1.0')
 build_coverage_matrix('dalli', [2])
 build_coverage_matrix('karafka', min: '2.3.0')
+build_coverage_matrix('waterdrop', min: '2.8.8.rc1')
```
Please run `bundle exec ruby appraisal/generate.rb && bundle exec rake dependency:lock['/app/gemfiles/ruby_3.*_waterdrop_*.gemfile']` for each supported Rubies so that the gemfiles for waterdrop are generated and used in the CI.

vpellan reviewed (commented):
Hello ! Thank you very much for your contribution ! I gave it a first pass and left a bunch of suggestions, feel free to pick them up but if you prefer, we can handle them !

*Not in comments: to fix lint errors, you can run `bundle exec rake standard:fix`
To generate prototype signature files, you can use `bundle exec rake rbs:prototype`
If at all possible, please try to change `untyped` items to the correct types! See the following guide: docs/StaticTypingGuide.md*

ericfirth committed:
Fix CI failures

dd-apm-ecosystems-autobot[bot] committed:
[🤖] Lock Dependency: https://github.com/DataDog/dd-trace-rb/actions/runs/19139423585

vpellan reviewed (approved):
LGTM! (Left two suggestions, I think the Matrixfile one should be applied before merging)

ericfirth committed:
Update Matrixfile

dd-apm-ecosystems-autobot[bot] committed:
[🤖] Lock Dependency: https://github.com/DataDog/dd-trace-rb/actions/runs/19336724142

dd-apm-ecosystems-autobot[bot] committed:
[🤖] Lock Dependency: https://github.com/DataDog/dd-trace-rb/actions/runs/19336947793

ericfirth committed:
Steep changes to ensure requisite typings for waterdrop files

ericfirth commented on Matrixfile:
```diff
@@ -127,6 +127,11 @@
+  'waterdrop' => {
+    # WaterDrop 2.8.8+ requires Ruby >= 3.1.0 (JRuby 9.4+ only, not 9.2/9.3)
+    'waterdrop-latest' => '❌ 2.5 / ❌ 2.6 / ❌ 2.7 / ❌ 3.0 / ✅ 3.1 / ✅ 3.2 / ✅ 3.3 / ✅ 3.4 / ✅ 3.5 / ✅ jruby-9.4+',
+    'waterdrop-min'    => '❌ 2.5 / ❌ 2.6 / ❌ 2.7 / ❌ 3.0 / ✅ 3.1 / ✅ 3.2 / ✅ 3.3 / ✅ 3.4 / ✅ 3.5 / ✅ jruby-9.4+'
```
fixed
